File: android/src/main/java/app/independo/capacitorvoicerecorder/adapters/RecorderAdapter.java
package app.independo.capacitorvoicerecorder.adapters;

import app.independo.capacitorvoicerecorder.core.CurrentRecordingStatus;
import app.independo.capacitorvoicerecorder.core.RecordOptions;
import app.independo.capacitorvoicerecorder.platform.NotSupportedOsVersion;
import java.io.File;
import java.util.function.Consumer;

/** Recorder abstraction used by the service layer. */
public interface RecorderAdapter {
    /** Sets a callback invoked when interruptions begin. */
    void setOnInterruptionBegan(Runnable callback);

    /** Sets a callback invoked when interruptions end. */
    void setOnInterruptionEnded(Runnable callback);

    /** Sets the callback for real-time volume updates. */
    void setOnVolumeChanged(Consumer<Float> callback);

    /** Starts recording audio. */
    void startRecording();

    /** Stops recording audio. */
    void stopRecording();

    /** Pauses recording if supported. */
    boolean pauseRecording() throws NotSupportedOsVersion;

    /** Resumes recording if supported. */
    boolean resumeRecording() throws NotSupportedOsVersion;

    /** Returns the current recording status. */
    CurrentRecordingStatus getCurrentStatus();

    /** Returns the output file for the recording. */
    File getOutputFile();

    /** Returns the options used to start recording. */
    RecordOptions getRecordOptions();

    /** Deletes the output file from disk. */
    boolean deleteOutputFile();
}

File: ios/Sources/VoiceRecorder/Adapters/RecorderAdapter.swift
import Foundation

/// Recorder abstraction used by the service layer.
protocol RecorderAdapter: AnyObject {
    /// Options supplied when recording starts.
    var options: RecordOptions? { get }
    /// Callback invoked when interruptions begin.
    var onInterruptionBegan: (() -> Void)? { get set }
    /// Callback invoked when interruptions end.
    var onInterruptionEnded: (() -> Void)? { get set }
    /// Callback for receiving volume updates.
    var onVolumeChanged: ((Float) -> Void)? { get set }

    /// Starts recording audio.
    func startRecording(recordOptions: RecordOptions?) -> Bool
    /// Stops recording and returns success via callback.
    func stopRecording(completion: @escaping (Bool) -> Void)
    /// Pauses recording if supported.
    func pauseRecording() -> Bool
    /// Resumes recording if supported.
    func resumeRecording() -> Bool
    /// Returns the current recording status.
    func getCurrentStatus() -> CurrentRecordingStatus
    /// Returns the output file for the current session.
    func getOutputFile() -> URL
}
